File: cs50x/psets/wk6/sentimental/sentimental-credit/credit.py
# Finished


def main():
    ccn = get_card_no()
    if ccn_validity(ccn):
        show_card_brand(ccn)
    else:
        print("INVALID")
        return 0


# Gets Credit Card Number from User
def get_card_no():
    while True:
        ccn = int(input("Card Number: " + "\n"))
        if ccn > 0:
            return ccn


# Evaluates the Length of Credit Card Number
def ccn_length(ccn):
    count = 0
    while ccn > 0:
        ccn = ccn // 10
        count = count + 1
    return count


# Evaluates the Brand of the Card
def show_card_brand(ccn):
    if (ccn >= 4e12 and ccn < 5e12) or (ccn >= 4e15 and ccn <= 5e15):
        print("VISA")
    elif (ccn >= 34e13 and ccn <= 35e13) or (ccn >= 37e13 and ccn <= 38e13):
        print("AMEX")
    elif (ccn >= 51e14) and (ccn < 56e14):
        print("MASTERCARD")
    else:
        print("INVALID")


# Validates the Credit Card Number using Luhn's Algortihm
def ccn_validity(ccn):
    i = 0
    _sum = 0
    digit = 0
    ccn_len = ccn_length(ccn)
    while ccn != 0:
        if i % 2 == 0:
            _sum += ccn % 10
        else:
            digit = (ccn % 10) * 2
            _sum += (digit % 10) + (digit // 10)
        i = i + 1
        ccn //= 10
    return (_sum % 10 == 0) and ((ccn_len) == 13 or (ccn_len) == 15 or (ccn_len) == 16)


if __name__ == "__main__":
    main()

File: cs50x/psets/wk1/credit/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

// Function declarations
int ccn_len(long ccn);
void Show_Card_Brand(long ccn);
bool ccn_validity(long ccn);

int main(void)
{
    // Get Credit Card Number from the User as input
    long ccn;
    do
    {
        ccn = get_long("Card Number: ");
    }
    while (ccn < 0);

    // Checks the validity of the Card Number
    if (ccn_validity(ccn))
    {
        Show_Card_Brand(ccn);
    }
    else
    {
        printf("INVALID\n");
        return 0;
    }
}

// Compute the Length of Card Number
int ccn_len(long ccn)
{
    int count;
    for (count = 0; ccn > 0; ccn /= 10, count++)
        ;
    return count;
}

// Finds the Brand of the Credit Card
void Show_Card_Brand(long ccn)
{
    if ((ccn >= 4e12 && ccn <= 5e12) || (ccn >= 4e15 && ccn <= 5e15))
    {
        printf("VISA\n");
    }
    else if ((ccn >= 34e13 && ccn <= 35e13) || (ccn >= 37e13 && ccn <= 38e13))
    {
        printf("AMEX\n");
    }
    else if (ccn >= 51e14 && ccn < 56e14)
    {
        printf("MASTERCARD\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

// Applies Luhn's Algorithm to Card Number
bool ccn_validity(long ccn)
{
    long ccn_tmp = ccn;
    int sum = 0;
    int digit = 0;

    for (int j = 0; ccn > 0; j++, ccn /= 10)
    {
        if (j % 2 == 0)
            sum += ccn % 10;
        else
        {
            digit = (ccn % 10) * 2;
            sum += (digit % 10) + (digit / 10);
        }
    }
    ccn = ccn_tmp;
    int card_len = ccn_len(ccn);
    return ((sum % 10 == 0) && ((card_len == 13) || (card_len == 15) || (card_len == 16)));
}
